### PR TITLE
feat: new metric widgets

### LINKED ISSUE #1769: [➕ Feature]: New widgets for Dashboards: MTTR, Alerts/hour, etc...
For now, dashboards support pretty basic preset-based widgets: Screenshot 2024-09-01 at 14 42 28

We need more widgets:

1.  MTTR (Mean Time To Resolve) — widget with a items/time graph, resolution: 24h.
2. Alerts / day — widget with a items/time graph, resolution: 24h.
3. Incidents / day — widget with a items/time graph, resolution: 24h.
4. Workflows / day — widget with a items/time graph, resolution: 24h.

### PR DESCRIPTION
**Pull request by 35C4n0r**
Closes #1769

## ✅ Checks

- [ ] My pull request adheres to the code style of this project
- [ ] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [ ] All the tests have passed

## ℹ Additional Information

https://www.loom.com/share/aa665dee5b2c413dac14976ed62e90fe?sid=00b8c786-a87a-47d7-a17f-a6dbe74de11f

### CODE REVIEW

**Review by Matvey-Kuk — APPROVED** [review]
Great job!

**Comment by rajeshj11** [comment]
@35C4n0r What if I want to see the previous day's data? how  are we handling it
cc: @Matvey-Kuk

**Comment by rajeshj11** [comment]
> @35C4n0r What if I want to see the previous/particular day's data? how are we handling it
@Matvey-Kuk User might ask for it, right? I want to know your perspective.
cc: @talboren

**Review by talboren — DISMISSED** [review]
@35C4n0r may I ask you to resolve the conflict, and run linter? For example, please note that every other component, the standard is using " instad of '.

May I also ask you to upload a video of how this feature looks like?

cc: @Matvey-Kuk - I know you approved but would love to see that ^ first.
re @rajeshj11: @Matvey-Kuk @35C4n0r I think it's a valid point. As a user I would like the dashboard to work with the timeframe I chose.

**Comment by 35C4n0r** [comment]
@talboren 

> @35C4n0r may I ask you to resolve the conflict, and run linter?

Conflict Resolved, is the command for linter `next lint` ?

> For example, please note that every other component, the standard is using " instad of '.

Didn't understand, can you clarify this.

> May I also ask you to upload a video of how this feature looks like?

It's there in the description
 
> cc: @Matvey-Kuk - I know you approved but would love to see that ^ first. re @rajeshj11: @Matvey-Kuk @35C4n0r I think it's a valid point. As a user I would like the dashboard to work with the timeframe I chose.

Ok, then instead of the past 24 hours, we'll show the graph for per day? Thoughts?

**Comment by 35C4n0r** [comment]
😅 
Suddenly clicked what you're trying to say. 

> For example, please note that every other component, the standard is using " instad of '.

**Comment by rajeshj11** [comment]
> @talboren
> 
> > @35C4n0r may I ask you to resolve the conflict, and run linter?
> 
> Conflict Resolved, is the command for linter `next lint` ?
> 
> > For example, please note that every other component, the standard is using " instad of '.
> 
> Didn't understand, can you clarify this.
> 
> > May I also ask you to upload a video of how this feature looks like?
> 
> It's there in the description
> 
> > cc: @Matvey-Kuk - I know you approved but would love to see that ^ first. re @rajeshj11: @Matvey-Kuk @35C4n0r I think it's a valid point. As a user I would like the dashboard to work with the timeframe I chose.
> 
> Ok, then instead of the past 24 hours, should we show the graph for per day? Thoughts?

@35C4n0r 
We can assume a similar approach. Pagination is necessary if the time frame is extensive to prevent overloading the component.
Screenshot from 2024-10-24 12-57-07

**Comment by 35C4n0r** [comment]
😓 @rajeshj11 that's definitely not what I meant. Let me rephrase it a bit:
Instead of showing the data for the past 24 hours starting from the current timestamp, we should display a graph that spans the selected day.

**Comment by rajeshj11** [comment]
> 😓 @rajeshj11 that's definitely not what I meant. Let me rephrase it a bit: Instead of showing the data for the past 24 hours starting from the current timestamp, we should display a graph that spans the selected day.

I understand your point, and I see it from that perspective as well. I’ve also noted in my comments that this is based purely on my assumption.

**Comment by rajeshj11** [comment]
@35C4n0r Let @Matvey-Kuk and @talboren decide which one they want. If you're open to it, I'd like to collaborate with you to tackle this blocker together.

File: keep/api/routes/incidents.py [hunk 1]
```diff
@@ -708,7 +710,10 @@ def change_incident_status(
 
     # We need to do something only if status really changed
     if not change.status == incident.status:
-        result = change_incident_status_by_id(tenant_id, incident_id, change.status)
+        end_time = (None, datetime.utcnow())[change.status == IncidentStatus.RESOLVED]
```
Comment (VladimirFilonov) [inline keep/api/routes/incidents.py:713]: It's hard to read such syntax, plus it includes automatic type casting. Let's stick to old-style 
```suggestion
        end_time = datetime.utcnow() if change.status == IncidentStatus.RESOLVED else None
```

**Review by VladimirFilonov — CHANGES_REQUESTED** [review]

File: keep/api/core/db.py [hunk 2]
```diff
@@ -2092,12 +2100,32 @@ def get_linked_providers(tenant_id: str) -> List[Tuple[str, str, datetime]]:
...
+def get_provider_distribution(
+    tenant_id: str,
+    aggregate_all: bool = False,
+    timestamp_filter: TimeStampFilter = None,
+) -> (
+    list[dict[str, int | Any]]
+    | dict[str, dict[str, datetime | list[dict[str, int]] | Any]]
+):
+    """
+    Returns hits per hour and the last alert timestamp for each provider in the past 24 hours.
```
Comment (Matvey-Kuk) [inline keep/api/core/db.py:2112]: I guess it's not 24h now? Since timestamp_filter is added.

**Review by Matvey-Kuk — APPROVED** [review]
It's looking really good for me. I see tests are failing, please fix them :)

**Comment by 35C4n0r** [comment]
> It's looking really good for me. I see tests are failing, please fix them :)

@Matvey-Kuk tests were failing sporadically, they all pass now. I've updated the docs. Lemme know if any other changes are required and let's get this PR merged :)

**Review by VladimirFilonov — APPROVED** [review]
I think it's great!
